### ISSUE TITLE
xEdit Cleaning produces different (broken) files if a cache is present

### ISSUE THREAD

**Issue by Umgak** [TES5Edit/TES5Edit#803]
**What is the version of xEdit you are using?**
SSEEdit 4.0.3 (1CDF3AC7)
SSEEdit 4.1.3a EXTREMELY EXPERIMENTAL (DF87E311)

**Describe the bug**
If a file adds worldspaces, and it is cleaned with xEdit, the cleaning operation will produce broken files if a refcache is loaded. These broken files have large portions of _some_ of their worldspace's chunks corrupted. For the Soul Cairn and Apocrypha, this is all chunks with X or Z coordinates less than 0. For Darkfall Passage, the coordinates are instead X<-36, Z>16 due to that region's strangely mapped coordinate space.

**To reproduce**
1. Delete SSEEdit Cache
2. Restore uncleaned file (either by reinstalling mod, restoring from backup, or verifying integrity of game cache)
3. Clean file with QuickAutoClean. This cleaned file should work as expected.
4. Restore uncleaned file again.
5. Re-clean file with QuickAutoClean

NOTE: Dawnguard.esm does not behave normally when cleaned, and thus must be manually cleaned using the old Apply Filter for Cleaning method to reproduce the bug.

**Expected behavior**
The produced file should be identical whether there is, or is not a cache, as the cache should ideally only be speeding up loading times and not modifying behaviour in any way.

**Screenshots**
DLC01SoulCairn in 0,0 facing toward -1,-1
![DLC01SoulCairn in 0,0 facing toward -1,-1](https://user-images.githubusercontent.com/40526330/77586084-8ce15380-6eb3-11ea-9161-7c6c13756cf9.png)

2 screenshots from DLC1DarkfallPassage (coordinate space is strange here)
![DLC1DarkfallPassage](https://user-images.githubusercontent.com/40526330/77586316-fa8d7f80-6eb3-11ea-86bd-71d3ebccf346.png)
![DLC1DarkfallPassage](https://user-images.githubusercontent.com/40526330/77586324-fcefd980-6eb3-11ea-8e8a-a20d31811f5e.png)

DLC2ApocryphaWorld in 0,0 facing toward -1,-1
![DLC2ApocryphaWorld in 0,0 facing toward -1,-1](https://user-images.githubusercontent.com/40526330/77585788-17758300-6eb3-11ea-8b6d-228e7b1d969a.png)

**Additional context**
A list of known hashes & sizes for working and broken files in Skyrim Special Edition is as follows:
4.0.3
Update.esm            EF461BD9, 18090304 bytes (Working)     UNAFFECTED
Dawnguard.esm      955CCA77, 24284359 bytes (Working)   C7F92E24, 24341725 bytes (Pass 1 manual cleaning, broken)     43655D8E, 24340112 bytes (Pass 2 manual cleaning, still broken)
HearthFires.esm      B384A732, 3681366 bytes   (Working)    UNAFFECTED
Dragonborn.esm     D8C00B93, 64166711 bytes (Working)   3E903FCB, 64357633 bytes (Broken)

4.1.3a
Update.esm            F1DD0ADC, 18090423 bytes (Working)    6E8CA420, 18090949 bytes (presumably damaged somehow, but the only added worldspace is a test area that doesn't break)
Dawnguard.esm      AD407311, 24284588 bytes (Working)   7EA3B18C, 24341954 bytes (Pass 1 manual cleaning, broken)     FA026978, 24340341 bytes (Pass 2 manual cleaning, still broken)
HearthFires.esm      D4095EC4, 3681454 bytes (Working)   E4B2D2C9, 3681436 bytes (Presumably damaged somehow)
Dragonborn.esm     6973A406, 64166519 bytes (Working)   72D9BB0D, 64357411 bytes (Broken)

**Comment by Sentrus** [comment]
I reproduced the same results as Umgak. Any update on this?

**Comment by ElminsterAU** [comment]
The issue is because the original files from Bethesda contain the block/sub-block groups in these worldspaces in the wrong order. I'm looking into writing an OFST subrecord in the worldspace so that the game can use that to find the CELLs instead of having to depend on the correct ordering of the block/sub-block groups.

**Comment by Umgak** [comment]
4.1.3b prevents this bug from manifesting when using QuickAutoClean function, but the same regions are still broken when manually cleaning. Note that 4.0.3 and 4.1.3a also could not cause this bug with Dawnguard when QuickAutoCleaning - now Dragonborn also works when QuickAutoCleaned even if cache is present, which is a definite improvement.

**Comment by fireundubh** [comment]
What are you referring to as "manual cleaning"?

**Comment by Umgak** [comment]
Apply Filter for Cleaning -> Undelete and Disable References/Remove "Identical to Master" records, which is the old method of cleaning pre-QuickAutoClean and can still be used by enabling it in the options. It is still _occasionally_ useful, but largely surpassed by QuickAutoClean function now.

**Comment by ElminsterAU** [comment]
He's doing what he isn't supposed to be doing, starting xEdit such that he can reactivate the old cleaning menu items in the experts tab and then use these. And frankly I don't give a rats behind about it. That is not a supported way to modify any of Bethesda broken DLC.

I know exactly what's going on. The original DLCs are structurally broken. They contain worldspaces with the block/sub-block groups in the wrong order. The runtime code in the engine depends on them being in a specific order to be able to find the temporary CELL records at to on-demand load their contents. Which then fails. 

The reason why it doesn't fail with the uncleaned files is because they contain an optional OFST subrecord in the WRLD record, which contains a direct mapping from cell grid to file offset for all CELLs contained in the worldspace, so that the engine can use that instead of the normal way to find the CELLs at runtime. 

These OFST records are optional. xEdit can't generate them (and neither can the CK). The only files that actually have them are the original game master and DLC from Bethesda. If xEdit where to just save the loaded contents of these OFST subrecords back into the files on save, it would result in corrupted files as the offsets are wrong for the newly saved file. So xEdit removes the OFST on load (to prevent it from being saved and crashing everything).

For performance reasons and to reduce memory consumption by orders of magnitude xEdit does not load and parse the contests of records when the program starts, and when saving, if a record or group is not marked as modified, then xEdit directly writes out the bytes from the original file again with having to reserialize them from individual elements.

That means that if the child groups of the broken worldspaces don't get modified, then they don't sort their contained block and sub-block groups and simply write out the complete child group by copying the bytes from the original file unchanged. (This saves a lot of time and memory.)

As mentioned above, when starting xEdit, it does not load/parse the contents of records and instead records are parsed only on demand when their contents is accessed. 

When xEdit is building references, it needs to look into the contents of every record which could contain FormIDs. This parsing/loading of records is what costs all the time during start when building references. 

As part of loading records, the AfterLoad callback attached to them is executed. Some of these are making changes to the records in "internal edit" mode after loading to correct certain data inconsistencies. "Internal edit" mode means that while the elements are modified (and notifications of this are bubbled upwards, marking all parents as modified, triggering sorting and so on), they don't show as bold modified to the user.

So, a side effect of building references is that all the records get loaded, their AfterLoad callback executed, at least one record somewhere in the broken worldspaces in the DLC gets internally edited, resulting in the worldspace child group being marked modified so on save it writes out the now correctly sorted block groups one by one instead of simply writing the unchanged bytes from the source file.

When the cache is used, the records don't have to be loaded, the after load callbacks don't get called, the worldspace child groups don't get marked as modified and on save the original bytes are written for them. 

Which isn't a problem, and in fact is much preferable, except for the DLC, which following longstanding Bethesda tradition are all, every single one of 'em, a horrifying broken mess.

QAC in 4.0.2 worked around this by calling mark modified recursively on all groups and records in the file before saving. Which is where this sage should have ended. 

Except that in Dragonborn, the NPC records are stored in such an order that the game engine can't resolve an attack template reference on load, and if that gets fixed by the mark modified, the engine can now load all the records correctly, which makes bristlebacks behaved non-aggresively, *as the record contents says they should*, instead if being aggressive because the game engine couldn't correctly load the records.

So 4.0.3 fixed this by now longer calling mark modified on every file, instead it has a list of files that need it, which only contained dawnguard (to fix soul cairn), me not knowing that they messed up Apocrypha as well.

4.1.3b now has both dawnguard and Dragonborn in that list (to fix soul cairn and apocrypha), but only calls mark modified recursively on the WRLD top level group instead of the whole file (so as not to fix, and thereby break, bristlebacks).

In summary, the DLC are all a horrifying broken mess that Bethesda somehow frankensteined together in a way that CK would never save it, and any modification to the them outside of the carefully controlled environment of QAC is not supported.

**Comment by Umgak** [comment]
Well, QAC works perfectly fine on 4.1.3b, so this can be considered resolved then. Curse you, Bethesda jank.
